### ISSUE TITLE
go-plus cannot find my gopath and no features work

### ISSUE THREAD

**Issue by Uglyface200** [joefitzgerald/go-plus#386]
Operating system: Xubuntu 14.04

Atom version: 1.6.0

When I start Atom with a Go file open, three dialog boxes appear. A fourth appears after I right-click on a variable and select "rename symbol under cursor".

![screenshot - 16-03-17 - 06 41 20 pm](https://cloud.githubusercontent.com/assets/2916723/13866178/1358a79e-ec70-11e5-97e2-34b0286a7d52.png)

When I click the "Run Go Get" button on each of the dialog boxes, they all become error messages:

![screenshot - 16-03-17 - 06 40 49 pm](https://cloud.githubusercontent.com/assets/2916723/13866204/3d4f0dc2-ec70-11e5-9803-00ea28327815.png)

When I first went through this process upon installing Atom, I manually installed all the required packages myself through the terminal. When I did, the command was pasted as `go get -ugolang.org/x/tools/cmd/goimports`. Note the lack of a space between the u and the URL; is this an issue? In any case, in my gopath, ~/go, the /bin directory now contains all the executables the dialog boxes require:

![screenshot - 16-03-17 - 06 46 32 pm](https://cloud.githubusercontent.com/assets/2916723/13866263/cfb48f48-ec70-11e5-84a7-67c23ffaf58f.png)

But the error still appears every time I start atom, and none of the features of go-plus work. The dialog boxes also appear in Atom Beta.

I went to the settings to configure the go-config package and feed it my gopath, but I cannot see any button to change its settings.

**Comment by joefitzgerald** [comment]
Yes, you need a space between -u and the github.... portion if you are installing these manually.

**Comment by atombender** [comment]
Same here, 4.0 broke everything for me — and I'm always launching Atom from the command line.

After installing all deps, linting and vetting stopped working, and format-on-save [moves the cursor on save](https://github.com/joefitzgerald/gofmt/issues/1).

There are errors in the console [about godef](https://github.com/litgh/atom-godef/issues/9), but that's all.

For some reason, `go-plus` no longer appears under "Installed Packages". To find its setting I have to go to "Install" and search there. Not sure if Atom bug or what.

**Comment by joefitzgerald** [comment]
The settings for go-plus are now in the individual packages. What operating system are you on, and what is the output of `process.env` in the Chromium console?

**Comment by atombender** [comment]
Interestingly, `process.env` does _not_ show my normal environment, even though I started Atom from the command line. `GOPATH` is not there. (Note that I have reverted to Atom 1.5 and go-plus 3.5.)

**Comment by intabulas** [comment]
Same issue for me with 1.6 and 4.0. I have a project in my gopath, that has sub packages and launching atom from the command line (in the root of the project) gives me linter errors not being able to find the sub packages.

I also have local vendored deps that it can't find in vendor/ either.. All worked just fine with 3.5 and 1.5x

**Comment by joefitzgerald** [comment]
Is someone open to doing a screen sharing session with me tomorrow on this? I very much want to see this myself.

**Comment by maximesong** [comment]
Same issue for me with 1.6 and 4.0. It seems to work again after I upgrade to atom 1.7 beta0

**Comment by mackjoner** [comment]
@maximesong thanks.. :-)

**Comment by pbnsilva** [comment]
Same issue here after update to 1.6.

**Comment by joefitzgerald** [comment]
So where do we stand on this issue? Is this still an issue?

**Comment by pbnsilva** [comment]
Definitely something got broken with update to atom 1.6 and go-plus 4.0.
After I started atom via terminal some of the issues disappeared, but I still get several errors that seem to be related with not being able to find GOPATH (can't find import).

**Comment by joefitzgerald** [comment]
@pbnsilva what operating system are you on? Is someone open to screen sharing with me so I can see this happening myself?

**Comment by pbnsilva** [comment]
@joefitzgerald i'm on a mac 10.11 .
i'm open to screen sharing, but only tomorrow

**Comment by jmelfi** [comment]
@joefitzgerald I'm on mac 10.11 and can verify that Atom 1.6.0 and go-plus 4.0 has this issue and didn't before.  I'd be open to doing a screenshare or something today or tomorrow if that would help.  Feel free to contact me here or by email.

**Comment by jmelfi** [comment]
@joefitzgerald I have verified that using the terminal 100% resolves my issue.  It is only when launching via the Applications folder or the Dock in Mac that causes the GOPATH variable to not be passed.

**Comment by atombender** [comment]
I was wrong about `process.env` before. Turns out Chromium will not print the whole thing and there's a tiny `…` before the last brace. I can confirm that `process.env.GOPATH` is correct.

**Comment by c2h5oh** [comment]
Ubuntu 15.10, Atom 1.6, go-plus 4.0 - same problem

`process.env` does not include GOPATH. `echo $GOPATH` in terminal returns a valid GOPATH.

**Comment by mackjoner** [comment]
System: Mac OS X 10.11
Atom: 1.6.0
go-plus: 4.0.1

terminal open atom
```
GOPATH={your GOPATH} {your atom app path}/Contents/MacOS/Atom
```
this for me.
```
GOPATH=/Users/yangzhaojie/local/gocode /Applications/Atom.app/Contents/MacOS/Atom
```

it's work good.

**Comment by intabulas** [comment]
Happy to screen share tomorrow.

**Comment by c2h5oh** [comment]
Interesting. I've just installed 1.7beta0 and.. it fixed 1.6.0. Fix is permanent - still works after system restart.

The thing is, at least on Linux, beta channel and stable channel share nothing except `~/.atom/` directory, so there should be interaction and yet there is. `~/.atom/config.cson` modification timestamp is roughly the same as 1.7.0beta0 installation timestamp - maybe there is the problem?

**Comment by Uglyface200** [comment]
For what it's worth, when I manually installed the package, I of course included a space between the -u and the url name. I said I didn't know whether it was relevant because it seemed highly probable that whatever was being copied to the clipboard was the command that was being run internally (and thus failed).

**Comment by bakoonin** [comment]
Had similar problems with go-plus 4.0.1 on W8.1 and Atom 1.6.x. Upgrading to Atom 1.7.0-beta0 fixed it.

**Comment by sandlerben** [comment]
Had the same issue with go-plus 4.0 and Atom 1.6. Atom 1.7.0-beta fixed it.

**Comment by joefitzgerald** [comment]
Please update `environment` to `v1.2.0` and retest if you are on OS X.

**Comment by pbnsilva** [comment]
@joefitzgerald updated and problems persist: several import errors reported by gotype.

**Comment by joefitzgerald** [comment]
@pbnsilva Please restart atom, and then provide the actual errors.

**Comment by pbnsilva** [comment]
@joefitzgerald , for example:
![image](https://cloud.githubusercontent.com/assets/2334434/13907402/94abc2fc-eeed-11e5-8d49-677cdfebd7bd.png)

**Comment by pierrre** [comment]
@pbnsilva its a gotype/gometalinter issue: https://github.com/alecthomas/gometalinter/issues/40

**Comment by pbnsilva** [comment]
@pierrre thanks. I've disabled gotype.

**Comment by pierrre** [comment]
@pbnsilva or you can install the package

**Comment by pbnsilva** [comment]
@pierrre that would be fine for 3rd party packages, but an hassle otherwise. I'd rather rely on different tools.

**Comment by atombender** [comment]
@joefitzgerald: After upgrading `environment`, it seems `go-plus` 4.0.1 is back to how 3.5 was. Great, thanks.

However, as an aside, I really question the set of default linters now used through `gometalinter`, and the fact that you have to fiddle with command line options to disable the most intrusive ones.

`gotype` is, as @pbnsilva discovered, apparently requires referenced packages to be _installed_ into `$GOPATH` first, or it will be complain. That's pointless to include in a linter — who does that while working on a file in an editor?

`dupl` also seems like an odd choice to include by default, since its output trends towards false positives.

I also disable `golint` since it makes a lot of stupid complaints about missing documentation comments, which interferes with my workflow, but that's more of a personal decision.

**Comment by Xiol** [comment]
Still having this issue on Atom 1.6.0 with environment at 1.2.0 as recommended. This is on Ubuntu 15.10 and I can confirm that Atom is seeing the correct GOPATH. My issue is exactly as described by the original poster.

**Comment by pbnsilva** [comment]
Solution for me was to uninstall go-plus and gometalinter and use other packages, separately.

**Comment by joefitzgerald** [comment]
@pbnsilva your issues are different to those being experienced by the Ubuntu users in this thread.

Regarding your decision to uninstall gometalinter-linter, in fact one of the reasons I moved from a monolithic package to single purpose packages was to allow people to disable specific packages to turn off behavior they do not want.

If you keep go-plus installed and disable the gometalinter-linter package then you will get the benefit of additional packages included in go-plus in the future. Or you can keep an eye out for new packages and do this yourself - both are perfectly valid approaches.

**Comment by pbnsilva** [comment]
@joefitzgerald the issues I had initially were the same as the original poster's and started with the updates, so seemed reasonable to assume they were to blame. What was go-plus using before for code analysis ?

**Comment by pbnsilva** [comment]
@joefitzgerald I initially had the same issues as the original poster's and started after the updates, so was reasonable to assume they were to blame.

go-plus automatically installs gometalinter. How do I disable this dependency?

**Comment by joefitzgerald** [comment]
@pbnsilva Open package settings, and click the disable button.

**Comment by joefitzgerald** [comment]
:wave: Linux folks, particularly those running Ubuntu. I believe I have reproduced your issues and _I think_ they have to do with where you are setting your GOPATH and PATH.

* `~/.profile`: If you set them here, $GOPATH will be set in Atom when you launch Atom _for the first time_ from the launcher in the toolbar
* `~/.bashrc` If you set them here, $GOPATH will be set in Atom when you launch Atom _for the first time_ from the terminal

Thus, to ensure GOPATH and PATH are set correctly regardless of how you launch Atom, you should put the following (or similar) in _both_ `~/.profile` and `~/.bashrc`:

```
export GOPATH=$HOME/work
export PATH=$GOPATH/bin:/usr/local/go/bin:$PATH
```

Obvious the above doesn't apply exactly if bash isn't your default shell, but I trust that if you're a user of a different shell, you can grok what I am saying above and apply the same principles to your particular environment.

**Comment by den-is** [comment]
> Thus, to ensure GOPATH and PATH are set correctly regardless of how you launch Atom, you should put the following (or similar) in both `~/.profile` and `~/.bashrc`

Strange solution.

Why can't we control go settings in atom go-plus settings?
My setup is:
- Fedora 23 x64
- Custom go 1.6 installation:
```bash
$ which go
/opt/go16/bin/go
```
zsh 5.1.1 - so i'm not using .bashrc or .profile. i do have go related records in .zshrc like:
```bash
export GOROOT=/opt/go16
export PATH=$PATH:$GOROOT/bin
export GOPATH=$HOME/projects/gohome
```

still have problem:
![Go-plus settings](http://s18.postimg.org/52244hiax/Screenshot_from_2016_03_25_10_13_33.png)
when pressing go-get in those pop-ups
![Go-plus error](http://s17.postimg.org/rt4xtle3z/Screenshot_from_2016_03_25_10_17_09.png)

**Comment by joefitzgerald** [comment]
Hi @den-is, sorry you're having issues. go-plus now requires your environment to be sane. There are too many edge cases to handle if the environment isn't sane.

One of the things I want to explore with @lee-dohm is whether we should be patching the environment on Linux in the way that we do for OSX. I will do some testing today to see if that is 1) feasible and 2) desirable. 

In essence, if both are true, we could have code that automatically launches your shell to get the environment you think you intend to use.

The counter argument to that would be that there is clearly a place to set your environment for GUI apps / apps launched by launcher; and the user should a) be aware of that and b) be capable of settings things differently for GUI apps than terminal apps.

**Comment by madranet** [comment]
Have you done something to fix this?

I've just run the updates again and they all installed. I didn't change anything else [I was already on go-plus 4,1,0 environment 1,2,0]. I ran `process.env` before and after re-trying the updates and `GOPATH` wasn't there the first time and was the second. The only thing I did in betweeen was open Atom's `config.json` file and make sure that it contained my GOPATH, which it did. 

```
"go-plus":
    goPath: "/Users/madra/claraithe/go"
```

It seems almost like opening the file 'blessed' it and allowed my GOPATH to be found. Oh well. Whatever the cause, I'm back in business again. 

Cheers!

BTW: Mine was the [parallel issue](https://github.com/joefitzgerald/autocomplete-go/issues/14) referenced above. Thought I'd post my reply in here to keep things tidier.

**Comment by joefitzgerald** [comment]
@madranet Yes, `environment` `v1.2.0` includes the same logic that is included in Atom >=1.7.0 to patch your environment if you open Atom from `Finder`, `Dock`, or `Spotlight`. Thus, whether you open it from the terminal or `Finder`, `Dock`, or `Spotlight`, you have a sane environment.

You can delete that config and it'll still work.

**Comment by jbreitbart** [comment]
For me putting the config in .profile and .bashrc is not working even though I am using bash.
```
$ cat .profile 
# ~/.profile: executed by the command interpreter for login shells.
# This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login
# exists.
# see /usr/share/doc/bash/examples/startup-files for examples.
# the files are located in the bash-doc package.

# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.

####### GO
#export GOROOT=$HOME/bin/go_appengine/goroot
export GOROOT=/usr/local/go
export GOPATH=$HOME/workspace/go
PATH=$GOROOT/bin:$PATH:$HOME/bin/go_appengine/goroot
#######


# if running bash
#if [ -n "$BASH_VERSION" ]; then
#    # include .bashrc if it exists
#    if [ -f "$HOME/.bashrc" ]; then
#	. "$HOME/.bashrc"
#    fi
#fi

# set PATH so it includes user's private bin if it exists
#if [ -d "$HOME/bin" ] ; then
#    PATH="$HOME/bin:$PATH"
#fi

#export PATH="$PATH:$HOME/.rvm/bin" # Add RVM to PATH for scripting

#[[ -s "$HOME/.rvm/scripts/rvm" ]] && source "$HOME/.rvm/scripts/rvm" # Load RVM into a shell session *as a function*
```
process.env does not contain $GOPATH.

**Comment by joefitzgerald** [comment]
Remember, for the most part `~/.profile` changes will require a reboot to be reflected correctly (there are other ways, but a reboot is the simplest to explain).

**Comment by joefitzgerald** [comment]
Also, try to avoid setting `GOROOT` if you can avoid it.

**Comment by joefitzgerald** [comment]
Finally, please ensure `$GOPATH/bin` is in your `PATH`. Sorry for the staccato responses.

**Comment by jbreitbart** [comment]
Got it. Looks like I did not restart the GUI session properly and the old .profile was still active.
Thanks for your patience :smile:

**Comment by joefitzgerald** [comment]
:+1: happy to help.

**Comment by narenaryan** [comment]
I solved the issue by making sure all the below check boxes are ticked

Here are points I observed
* Always make sure you are working on /usr/local/go/bin 
* Make sure GOPATH is in both ~/.zshrc and ~/.profile
i.e these commands should be there in your ~.zshrc (or) ~/.bashrc and ~/.profile

    export GOPATH=/home/naren/mygo
   export PATH=$PATH::/usr/local/go/bin
   export PATH=$PATH:$GOPATH/bin`

Now install go-plus from package installer in Atom. Everything will work fine.

**Comment by vith** [comment]
Just ran into *another* way this can happen: processes started by `nautilus` (aka GNOME Files) were not receiving environment variables that were set in `~/.profile` until I added `dbus-update-activation-environment --systemd --all` to the end of `~/.profile` (the `--systemd` flag may be unnecessary for this particular situation).

Opening files with their default handler, right clicking and choosing "Open With...", or right clicking a folder and choosing "Open in Terminal" all had this problem.

Starting a terminal directly from GNOME Shell and running `atom` was working fine, as was starting Atom directly from GNOME Shell.

This affects at least GNOME 3.20 on Arch Linux.

This is of course not `go-plus`'s fault, but I suspect others may end up here with the same problem.

**Comment by kekeoki** [comment]
# SYSTEM INFORMATION
- Ubuntu 16 64 bit 
# ATOM INFORMATION
- Atom    : 1.8.0
- Electron: 0.36.8
- Chrome  : 47.0.2526.110
- Node    : 5.1.1

# ERROR
package github.com/nsf/gocode: cannot download, $GOPATH not set. For more details see: go help gopath

# echo $GOPATH
/home/sammy/work



Really would love to get all features out of this great plugin, but running goimport never works. running go get, go install, and all other go commands work great from command line.

in the meantime I have run all the go gets in terminal and things are working.

let me know if you guys can help or if I can provide more information,
Thanks guys.

**Comment by jasonwbarnett** [comment]
I believe I can re-produce this on demand, hope this helps.

I believe that if the google app engine go sdk is in PATH before go (/usr/local/bin/go) then it creates all sorts of problems.

**Comment by sbucek** [comment]
Same problem. When running from command line GOROOT is not found anymore....

**Comment by dmitrijs-balcers** [comment]
Fixed on OSX Sierra.
https://gist.github.com/vsouza/77e6b20520d07652ed7d
Instead of `~/.bashrc` I edited `~/.bash_profile`

**Comment by matuskacmar1** [comment]
![image](https://cloud.githubusercontent.com/assets/19708626/20298429/22012022-ab16-11e6-9b3e-ad8faeaecec7.png)
Mac OSX Sierra 10.12
Golang v1.7
Atom v1.12.2

This happens sometimes when I quit and reopen the atom. I would say once in 3 attempts. I'm launching atom trough icon not trough terminal.

**Comment by phantasm66** [comment]
Anyone experiencing this... the following will fix the issue:
`launchctl setenv GOPATH /Users/<username>/<your go workspace directory>`

**Comment by pathtrk** [comment]
I already have set $GOPATH, So in my case

```
launchctl setenv GOPATH $GOPATH
```

fixed the issue.
I can launch Atom through icon without error now.

**Comment by joefitzgerald** [comment]
This is related to some changes that I made to the core of Atom. An updated version of go-plus is forthcoming. In the interim, please launch atom from the terminal instead of via the dock/finder/spotlight.

**Comment by deitch** [comment]
@joefitzgerald so this is back? Can you confirm the version of go-plus in which it will be/is fixed, so we can confirm?

**Comment by joefitzgerald** [comment]
This is fixed again in v5.0.0 of go-plus.

**Comment by deitch** [comment]
@joefitzgerald hats off! I am updating right now.

**Comment by deitch** [comment]
Definitely works, @joefitzgerald thank you!

I had to update, restart atom, it then complained again about not finding them, but ran some internal updates, added some packages, one more restart, and then all ran smoothly.

**Comment by joefitzgerald** [comment]
Good to hear, thanks @dietch for validating!

**Comment by muesli** [comment]
Experiencing this again on latest Arch. Setting the GOPATH in the go-plus settings doesn't help either. Starting from terminal works.

**Comment by joefitzgerald** [comment]
@muesli what version of Atom and go-plus are you experiencing this with?

**Comment by muesli** [comment]
Atom 1.12.9
go-plus 5.1.0

**Comment by joefitzgerald** [comment]
Are you using a version of Atom downloaded from atom.io or are you using an unofficial distribution from webupd8team or similar?

**Comment by muesli** [comment]
It's the community/atom package from ArchLinux.

Also worth noting, that setting a manual GOPATH in go-plus config makes some of the parsing work, but tests e.g. still don't work. In that case the "Go" tab showing all the ENV vars also just shows "empty".

**Comment by joefitzgerald** [comment]
https://github.com/atom/design-decisions/blob/master/unofficial-atom-distributions.md

TL;DR... please download the official distribution and use that. If the issue is present with that, let me know.

**Comment by muesli** [comment]
Happy to try it out, even though this problem seems to have just started appearing with go-plus 5.1.0. No Atom updates were installed in the mean time.

**Comment by pierrre** [comment]
@muesli update your Atom
I've got the same environment, no issue for me

**Comment by muesli** [comment]
@pierrre As I've mentioned, I'm on latest atom (now in th meantime 1.13) and go-plus releases, and it's still happening here.

**Comment by muesli** [comment]
@joefitzgerald Tried the binary releases. Same behavior / issue though.
